### DISCUSSION TITLE
feature: add c_sharp to lazyvim.plugins.extras.lang

### DISCUSSION THREAD

**Discussion by JustALawnGnome7** [LazyVim/LazyVim#1102] [Ideas]
### Did you check the docs?

- [X] I have read all the LazyVim docs

### Is your feature request related to a problem? Please describe.

There are multiple issues that prevent the C# language server (OmniSharp) from working out-of-the-box:

https://github.com/OmniSharp/omnisharp-roslyn/issues/2238
https://github.com/OmniSharp/omnisharp-roslyn/issues/2483

### Describe the solution you'd like

Some user configurations have essentially become standard for developing in C# with Neovim, as well plugins that have been written to close the gaps left by OmniSharp's non-conformance to the LSP specification. I would like a `c_sharp.lua` file added to `lazyvim.plugins.extras.lang` to better allow LazyVim to work out-of-the-box with OmniSharp.

### Describe alternatives you've considered

N/A

### Additional context

Using other Lua files from `lazyvim.plugins.extras.lang` as a guide, the following is some of what I've managed to come up with. Its a somewhat hacky approach that at least gets around the issues mentioned above:

``` lua
return {

  -- add C# to treesitter
  {
    "nvim-treesitter/nvim-treesitter",
    opts = function(_, opts)
      if type(opts.ensure_installed) == "table" then
        vim.list_extend(opts.ensure_installed, { "c_sharp", })
      end
    end,
  },

  -- setup lspconfig for C#
  {
    "neovim/nvim-lspconfig",
    dependencies = {
      { "Hoffs/omnisharp-extended-lsp.nvim" },
      {
        -- configure omnisharp-vim to use the omnisharp binary from mason
        "OmniSharp/omnisharp-vim",
        config = function()
          vim.cmd([[
            let g:OmniSharp_server_path = "/home/user/.local/share/nvim/mason/bin/omnisharp"
          ]])
        end
      },
    },
    opts = {
      -- make sure mason installs the server
      servers = {
        ---@type lspconfig.options.omnisharp
        omnisharp = {
          keys = {
            -- override default LazyVim functionality with omnisharp-vim features
            { "gd", "<cmd>OmniSharpGotoDefinition<CR>", desc = "Goto Definition" },
            { "gy", "<cmd>OmniSharpGotoTypeDefinition<CR>", desc = "Goto Type Definition" },
          },
        }
      },
      setup = {
        omnisharp = function(_, opts)
          -- https://github.com/OmniSharp/omnisharp-roslyn/issues/2238
          opts.handlers = { ["textDocument/definition"] = require("omnisharp_extended").handler, }

          -- https://github.com/OmniSharp/omnisharp-roslyn/issues/2483
          opts.on_attach = function (client, _)
            local function toSnakeCase(str)
              return string.gsub(str, "%s*[- ]%s*", "_")
            end

            local tokenModifiers = client.server_capabilities.semanticTokensProvider.legend.tokenModifiers
            for i, v in ipairs(tokenModifiers) do
              tokenModifiers[i] = toSnakeCase(v)
            end

            local tokenTypes = client.server_capabilities.semanticTokensProvider.legend.tokenTypes
            for i, v in ipairs(tokenTypes) do
              tokenTypes[i] = toSnakeCase(v)
            end
          end
        end,
      }
    },
  },
```

**Comment by amaanq** [comment]
I don't really think omnisharp-vim & omnisharp-extended-lsp.nvim is necessary, lsp protocol provides the ability to goto definitions/type definitions and they are mapped by default in lazyvim.

**Comment by JustALawnGnome7** [comment]
If you want to use “Goto Definition” for BCL classes and functions (e.g. `System.Console`), they’re necessary as far as I can tell. Please see the issues I referenced above for more information.

**Comment by JustALawnGnome7** [comment]
UPDATE: this week's release of OmniSharp 1.39.8 has rendered the temporary hack for the following issue unnecessary: https://github.com/OmniSharp/omnisharp-roslyn/issues/2483.
